bootstrap/kubeadm/api/v1alpha3/kubeadmconfig_webhook.go

```diff
@@ -18,10 +18,29 @@ package v1alpha3
 
 import (
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
-func (r *KubeadmConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+const (
+	DefaultEtcdDataDir = "/var/lib/etcd"
+)
+
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-bootstrap-cluster-x-k8s-io-v1alpha3-kubeadmconfig,mutating=true,failurePolicy=fail,groups=bootstrap.cluster.x-k8s.io,resources=kubeadmconfigs,versions=v1alpha3,name=default.bootstrap.cluster.x-k8s.io
+
+var _ webhook.Defaulter = &KubeadmConfig{}
+
+func (k *KubeadmConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
+		For(k).
 		Complete()
 }
+
+func (k *KubeadmConfig) Default() {
+	if k.Spec.ClusterConfiguration != nil {
+		if k.Spec.ClusterConfiguration.Etcd.Local != nil {
+			if k.Spec.ClusterConfiguration.Etcd.Local.DataDir == "" {
```

Comment:
Doesn't kubeadm already do this?  https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/types/v1beta1/types.go#L279

The webhook is fine, but we should probably change the type to add optional and omitempty instead?

Reply:
I will add omitempty. And yep, kubeadm will default it for you. The only benefit of doing it here is that the default value will show up in the CAPI object. If we don't default here then we assume the user will know the kubeadm defaults, which seems fine since it's documented on the type. Will remove the web hook & test and reduce this to an API change

Reply:
Can we do the same for the other defaulted fields? Not in this PR, but I like the point you're making about observability

Reply:
Yep, I can open an issue